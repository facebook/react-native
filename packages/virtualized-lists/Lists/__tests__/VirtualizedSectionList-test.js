/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 * @oncall react_native
 */

'use strict';

import nullthrows from 'nullthrows';

const VirtualizedSectionList = require('../VirtualizedSectionList');
const React = require('react');
const ReactTestRenderer = require('react-test-renderer');

describe('VirtualizedSectionList', () => {
  it('renders simple list', async () => {
    let component;
    await ReactTestRenderer.act(() => {
      component = ReactTestRenderer.create(
        <VirtualizedSectionList
          sections={[
            // $FlowFixMe[incompatible-type]
            {title: 's1', data: [{key: 'i1'}, {key: 'i2'}, {key: 'i3'}]},
          ]}
          // $FlowFixMe[missing-local-annot]
          renderItem={({item}) => <item value={item.key} />}
          getItem={(data, key) => data[key]}
          getItemCount={data => data.length}
        />,
      );
    });
    expect(component).toMatchSnapshot();
  });

  it('renders empty list', async () => {
    let component;
    await ReactTestRenderer.act(() => {
      component = ReactTestRenderer.create(
        <VirtualizedSectionList
          sections={[]}
          renderItem={({item}) => <item value={item.key} />}
          getItem={(data, key) => data[key]}
          getItemCount={data => data.length}
        />,
      );
    });
    expect(component).toMatchSnapshot();
  });

  it('renders empty list with empty component', async () => {
    let component;
    await ReactTestRenderer.act(() => {
      component = ReactTestRenderer.create(
        <VirtualizedSectionList
          sections={[]}
          ListEmptyComponent={() => <empty />}
          ListFooterComponent={() => <footer />}
          ListHeaderComponent={() => <header />}
          getItem={(data, key) => data[key]}
          getItemCount={data => data.length}
          renderItem={({item}) => <item value={item.key} />}
        />,
      );
    });
    expect(component).toMatchSnapshot();
  });

  it('renders list with empty component', async () => {
    let component;
    await ReactTestRenderer.act(() => {
      component = ReactTestRenderer.create(
        <VirtualizedSectionList
          // $FlowFixMe[incompatible-type]
          sections={[{title: 's1', data: [{key: 'hello'}]}]}
          ListEmptyComponent={() => <empty />}
          getItem={(data, key) => data[key]}
          getItemCount={data => data.length}
          renderItem={({item}) => <item value={item.key} />}
        />,
      );
    });
    expect(component).toMatchSnapshot();
  });

  it('renders all the bells and whistles', async () => {
    let component;
    await ReactTestRenderer.act(() => {
      component = ReactTestRenderer.create(
        <VirtualizedSectionList
          ItemSeparatorComponent={() => <separator />}
          ListEmptyComponent={() => <empty />}
          ListFooterComponent={() => <footer />}
          ListHeaderComponent={() => <header />}
          sections={[
            // $FlowFixMe[incompatible-type]
            {
              title: 's1',
              // $FlowFixMe[incompatible-call]
              data: new Array<void>(5)
                .fill()
                .map((_, ii) => ({id: String(ii)})),
            },
          ]}
          getItem={(data, key) => data[key]}
          getItemCount={data => data.length}
          getItemLayout={({index}) => ({
            index: -1,
            length: 50,
            offset: index * 50,
          })}
          inverted={true}
          keyExtractor={(item, index) => item.id}
          onRefresh={jest.fn()}
          refreshing={false}
          renderItem={({item}) => <item value={item.id} />}
        />,
      );
    });
    expect(component).toMatchSnapshot();
  });

  it('handles separators correctly', async () => {
    const infos = [];
    let component;
    await ReactTestRenderer.act(() => {
      component = ReactTestRenderer.create(
        <VirtualizedSectionList
          ItemSeparatorComponent={props => <separator {...props} />}
          sections={[
            // $FlowFixMe[incompatible-type]
            {title: 's0', data: [{key: 'i0'}, {key: 'i1'}, {key: 'i2'}]},
          ]}
          renderItem={info => {
            infos.push(info);
            return <item title={info.item.key} />;
          }}
          getItem={(data, key) => data[key]}
          getItemCount={data => data.length}
        />,
      );
    });
    expect(component).toMatchSnapshot();

    ReactTestRenderer.act(() => {
      infos[1].separators.highlight();
    });
    expect(component).toMatchSnapshot();
    ReactTestRenderer.act(() => {
      infos[2].separators.updateProps('leading', {press: true});
    });
    expect(component).toMatchSnapshot();
    ReactTestRenderer.act(() => {
      infos[1].separators.unhighlight();
    });
    expect(component).toMatchSnapshot();
  });

  it('handles nested lists', async () => {
    let component;
    await ReactTestRenderer.act(() => {
      component = ReactTestRenderer.create(
        <VirtualizedSectionList
          // $FlowFixMe[incompatible-type]
          sections={[
            {title: 'outer', data: [{key: 'outer0'}, {key: 'outer1'}]},
          ]}
          renderItem={outerInfo => (
            <VirtualizedSectionList
              sections={[
                // $FlowFixMe[incompatible-type]
                {
                  title: 'inner',
                  data: [
                    {key: outerInfo.item.key + ':inner0'},
                    {key: outerInfo.item.key + ':inner1'},
                  ],
                },
              ]}
              horizontal={outerInfo.item.key === 'outer1'}
              renderItem={innerInfo => {
                return <item title={innerInfo.item.key} />;
              }}
              getItem={(data, key) => data[key]}
              getItemCount={data => data.length}
            />
          )}
          getItem={(data, key) => data[key]}
          getItemCount={data => data.length}
        />,
      );
    });
    expect(component).toMatchSnapshot();
  });

  describe('scrollToLocation', () => {
    const ITEM_HEIGHT = 100;

    const createVirtualizedSectionList = async (props?: {
      stickySectionHeadersEnabled: boolean,
    }) => {
      let component;
      await ReactTestRenderer.act(() => {
        component = ReactTestRenderer.create(
          <VirtualizedSectionList
            sections={[
              // $FlowFixMe[incompatible-type]
              {
                title: 's1',
                data: [{key: 'i1.1'}, {key: 'i1.2'}, {key: 'i1.3'}],
              },
              // $FlowFixMe[incompatible-type]
              {
                title: 's2',
                data: [{key: 'i2.1'}, {key: 'i2.2'}, {key: 'i2.3'}],
              },
            ]}
            renderItem={({item}) => <item value={item.key} />}
            getItem={(data, key) => data[key]}
            getItemCount={data => data.length}
            getItemLayout={(data, index) => ({
              length: ITEM_HEIGHT,
              offset: ITEM_HEIGHT * index,
              index,
            })}
            {...props}
          />,
        );
      });

      const instance = nullthrows(component).getInstance();
      const spy = jest.fn();

      // $FlowFixMe[incompatible-use] wrong types
      // $FlowFixMe[prop-missing] wrong types
      instance._listRef.scrollToIndex = spy;

      return {
        instance,
        spy,
      };
    };

    it('when sticky stickySectionHeadersEnabled={true}, header height is added to the developer-provided viewOffset', async () => {
      const {instance, spy} = await createVirtualizedSectionList({
        stickySectionHeadersEnabled: true,
      });

      const viewOffset = 25;

      // $FlowFixMe[prop-missing] scrollToLocation isn't on instance
      instance?.scrollToLocation({
        sectionIndex: 0,
        itemIndex: 1,
        viewOffset,
      });
      expect(spy).toHaveBeenCalledWith({
        index: 1,
        itemIndex: 1,
        sectionIndex: 0,
        viewOffset: viewOffset + ITEM_HEIGHT,
      });
    });

    it.each([
      [
        // prevents #18098
        {sectionIndex: 0, itemIndex: 0},
        {
          index: 0,
          itemIndex: 0,
          sectionIndex: 0,
          viewOffset: 0,
        },
      ],
      [
        {sectionIndex: 2, itemIndex: 1},
        {
          index: 11,
          itemIndex: 1,
          sectionIndex: 2,
          viewOffset: 0,
        },
      ],
      [
        {
          sectionIndex: 0,
          itemIndex: 1,
          viewOffset: 25,
        },
        {
          index: 1,
          itemIndex: 1,
          sectionIndex: 0,
          viewOffset: 25,
        },
      ],
    ])(
      'given sectionIndex, itemIndex and viewOffset, scrollToIndex is called with correct params',
      async (scrollToLocationParams, expected) => {
        const {instance, spy} = await createVirtualizedSectionList();
        // $FlowFixMe[prop-missing] scrollToLocation not on instance
        instance?.scrollToLocation(scrollToLocationParams);
        expect(spy).toHaveBeenCalledWith(expected);
      },
    );
  });
});
