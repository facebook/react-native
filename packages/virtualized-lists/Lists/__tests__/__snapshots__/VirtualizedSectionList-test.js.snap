// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualizedSectionList handles nested lists 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "key": "outer0",
          },
          Object {
            "key": "outer1",
          },
        ],
        "title": "outer",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onStickyHeaderLayout={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={Array []}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        data={
          Array [
            Object {
              "data": Array [
                Object {
                  "key": "outer0:inner0",
                },
                Object {
                  "key": "outer0:inner1",
                },
              ],
              "title": "inner",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        keyExtractor={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        onStickyHeaderLayout={[Function]}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={Array []}
      >
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        />
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <item
            title="outer0:inner0"
          />
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <item
            title="outer0:inner1"
          />
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        />
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <RCTScrollView
        data={
          Array [
            Object {
              "data": Array [
                Object {
                  "key": "outer1:inner0",
                },
                Object {
                  "key": "outer1:inner1",
                },
              ],
              "title": "inner",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={true}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        onStickyHeaderLayout={[Function]}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={Array []}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          />
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <item
              title="outer1:inner0"
            />
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <item
              title="outer1:inner1"
            />
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          />
        </View>
      </RCTScrollView>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;

exports[`VirtualizedSectionList handles separators correctly 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "key": "i0",
          },
          Object {
            "key": "i1",
          },
          Object {
            "key": "i2",
          },
        ],
        "title": "s0",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onStickyHeaderLayout={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={Array []}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i0"
      />
      <separator
        highlighted={false}
        leadingItem={
          Object {
            "key": "i0",
          }
        }
        section={
          Object {
            "data": Array [
              Object {
                "key": "i0",
              },
              Object {
                "key": "i1",
              },
              Object {
                "key": "i2",
              },
            ],
            "title": "s0",
          }
        }
        trailingItem={
          Object {
            "key": "i1",
          }
        }
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i1"
      />
      <separator
        highlighted={false}
        leadingItem={
          Object {
            "key": "i1",
          }
        }
        section={
          Object {
            "data": Array [
              Object {
                "key": "i0",
              },
              Object {
                "key": "i1",
              },
              Object {
                "key": "i2",
              },
            ],
            "title": "s0",
          }
        }
        trailingItem={
          Object {
            "key": "i2",
          }
        }
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i2"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;

exports[`VirtualizedSectionList handles separators correctly 2`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "key": "i0",
          },
          Object {
            "key": "i1",
          },
          Object {
            "key": "i2",
          },
        ],
        "title": "s0",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onStickyHeaderLayout={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={Array []}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i0"
      />
      <separator
        highlighted={true}
        leadingItem={
          Object {
            "key": "i0",
          }
        }
        section={
          Object {
            "data": Array [
              Object {
                "key": "i0",
              },
              Object {
                "key": "i1",
              },
              Object {
                "key": "i2",
              },
            ],
            "title": "s0",
          }
        }
        trailingItem={
          Object {
            "key": "i1",
          }
        }
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i1"
      />
      <separator
        highlighted={true}
        leadingItem={
          Object {
            "key": "i1",
          }
        }
        section={
          Object {
            "data": Array [
              Object {
                "key": "i0",
              },
              Object {
                "key": "i1",
              },
              Object {
                "key": "i2",
              },
            ],
            "title": "s0",
          }
        }
        trailingItem={
          Object {
            "key": "i2",
          }
        }
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i2"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;

exports[`VirtualizedSectionList handles separators correctly 3`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "key": "i0",
          },
          Object {
            "key": "i1",
          },
          Object {
            "key": "i2",
          },
        ],
        "title": "s0",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onStickyHeaderLayout={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={Array []}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i0"
      />
      <separator
        highlighted={true}
        leadingItem={
          Object {
            "key": "i0",
          }
        }
        section={
          Object {
            "data": Array [
              Object {
                "key": "i0",
              },
              Object {
                "key": "i1",
              },
              Object {
                "key": "i2",
              },
            ],
            "title": "s0",
          }
        }
        trailingItem={
          Object {
            "key": "i1",
          }
        }
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i1"
      />
      <separator
        highlighted={true}
        leadingItem={
          Object {
            "key": "i1",
          }
        }
        press={true}
        section={
          Object {
            "data": Array [
              Object {
                "key": "i0",
              },
              Object {
                "key": "i1",
              },
              Object {
                "key": "i2",
              },
            ],
            "title": "s0",
          }
        }
        trailingItem={
          Object {
            "key": "i2",
          }
        }
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i2"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;

exports[`VirtualizedSectionList handles separators correctly 4`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "key": "i0",
          },
          Object {
            "key": "i1",
          },
          Object {
            "key": "i2",
          },
        ],
        "title": "s0",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onStickyHeaderLayout={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={Array []}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i0"
      />
      <separator
        highlighted={false}
        leadingItem={
          Object {
            "key": "i0",
          }
        }
        section={
          Object {
            "data": Array [
              Object {
                "key": "i0",
              },
              Object {
                "key": "i1",
              },
              Object {
                "key": "i2",
              },
            ],
            "title": "s0",
          }
        }
        trailingItem={
          Object {
            "key": "i1",
          }
        }
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i1"
      />
      <separator
        highlighted={false}
        leadingItem={
          Object {
            "key": "i1",
          }
        }
        press={true}
        section={
          Object {
            "data": Array [
              Object {
                "key": "i0",
              },
              Object {
                "key": "i1",
              },
              Object {
                "key": "i2",
              },
            ],
            "title": "s0",
          }
        }
        trailingItem={
          Object {
            "key": "i2",
          }
        }
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        title="i2"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;

exports[`VirtualizedSectionList renders all the bells and whistles 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  ListFooterComponent={[Function]}
  ListHeaderComponent={[Function]}
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "id": "0",
          },
          Object {
            "id": "1",
          },
          Object {
            "id": "2",
          },
          Object {
            "id": "3",
          },
          Object {
            "id": "4",
          },
        ],
        "title": "s1",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  invertStickyHeaders={true}
  inverted={true}
  isInvertedVirtualizedList={true}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[MockFunction]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onStickyHeaderLayout={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[MockFunction]}
      refreshing={false}
    />
  }
  refreshing={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "transform": Array [
          Object {
            "scaleY": -1,
          },
        ],
      },
      undefined,
    ]
  }
>
  <RCTRefreshControl />
  <View>
    <View
      collapsable={false}
      onLayout={[Function]}
      style={
        Object {
          "transform": Array [
            Object {
              "scaleY": -1,
            },
          ],
        }
      }
    >
      <header />
    </View>
    <View
      onFocusCapture={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    />
    <View
      onFocusCapture={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <separator />
      <item
        value="0"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <separator />
      <item
        value="1"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <separator />
      <item
        value="2"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <separator />
      <item
        value="3"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <item
        value="4"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    />
    <View
      onLayout={[Function]}
      style={
        Object {
          "transform": Array [
            Object {
              "scaleY": -1,
            },
          ],
        }
      }
    >
      <footer />
    </View>
  </View>
</RCTScrollView>
`;

exports[`VirtualizedSectionList renders empty list 1`] = `
<RCTScrollView
  data={Array []}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onStickyHeaderLayout={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={Array []}
>
  <View />
</RCTScrollView>
`;

exports[`VirtualizedSectionList renders empty list with empty component 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  ListFooterComponent={[Function]}
  ListHeaderComponent={[Function]}
  data={Array []}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onStickyHeaderLayout={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={Array []}
>
  <View>
    <View
      collapsable={false}
      onLayout={[Function]}
    >
      <header />
    </View>
    <empty />
    <View
      onLayout={[Function]}
    >
      <footer />
    </View>
  </View>
</RCTScrollView>
`;

exports[`VirtualizedSectionList renders list with empty component 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "key": "hello",
          },
        ],
        "title": "s1",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onStickyHeaderLayout={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={Array []}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        value="hello"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;

exports[`VirtualizedSectionList renders simple list 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "key": "i1",
          },
          Object {
            "key": "i2",
          },
          Object {
            "key": "i3",
          },
        ],
        "title": "s1",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onStickyHeaderLayout={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={Array []}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        value="i1"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        value="i2"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <item
        value="i3"
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;
